88.md

```diff
@@ -0,0 +1,107 @@
+NIP-88
+======
+
+Binary event encoding
+---------------------
+
+`draft` `optional` `author:dr-orlovsky`
+
+At the present moment nostr (NIP-1) uses text serialization of the event data,
+represented in a form of a JSON object. While being simple-to-implement and 
+web-compatible, it has some significant drawbacks:
+
+* Speed: serialization from JSON consumes most of the app time, according to [1]
+* Complex handling of binary data, which have to be encoded into Base64 or some
+  other binary-to-ascii format, increasing the problem described above and 
+  leading to more traffic consumption.
+* No explicit limits for the size of the data (i.e. it is not clear what is the
+  maximal size of the message, or the maximal number of tags), leading to DoS 
+  vulnerability and interoperability. While each relay is free to choose the
+  limits to prevent DoS attacks, the limits are not yet there, opening an attack
+  surface which may down the network for some time. Also, absence of the 
+  standard will lead that some of the relays will stop propagating some events 
+  (as a part of DoS prevention), while others will support all of them, making 
+  it hard to get the predictable and consistent user experience.
+* Protocol extensibility: with the current version of the protocol it is 
+  prohibited to add new fields to the JSON format.
+
+These all problems can be solved with introducing binary encoding to the 
+messages with strict limits to the maximal size of the elements and optional TLV
+extension stream.
+
+Specification
+-------------
+
+The proposed binary encoding must follow this spec (in pseudocode):
+```haskell
+Event ::
+    -- fields matching NIP-1:
```

Comment:
Any reason not to start with a one-byte version?

Reply:
TLVs. With just a version you either support it - or not - in full. With TLVs a gradual improvements are possible.

The only thing which version does (and TLVs can't) is signalling of the overall encoding standard. However, it can't do much even here: you can't migrate text/binary encodings with version byte (since it can't be present in the text encoding); it can't be use to migrate to protobufs or other common standard (since they prefix all fields with their own flags). So I do not see any use for the version byte here.

Reply:
> since it can't be present in the text encoding)

Is the text encoding meant to be "the default" (or at least a standard option) forever, or only kept until ecosystem has a time to migrate?


> The only thing which version does (and TLVs can't) is signalling of the overall encoding standard.

I guess I see your point. Both T in TLV and possibly even Kind, could act as a "version" But then ... wouldn't you want NIP-1 fields to be enclosed in a TLV?

I'm thinking - what if eventually it seems desirable to change NIP-1 fields? Right now there's no way to obsolete them. If they lived in their own TLV, clients can start creating events with one TLV for NIP-1, and another with whatever new format.

Reply:
I think we can close the topic since this PR is was concept NACKed. But let me answer just of scientific interest.

Yes, I was thinking about making all fields TLVs - but this makes base implementation more complex without any win. Even if all the fields are TLVs, you can't eventually get rid of them, since there is no rule in TLV saying "an required field may become obsolete and not required".

Reply:
Not each. All NIP-1 fields become one TLV with some "NIP-1" type. Then you eventually drop them all, as nothing is supposed to use them anymore.